src/lazy/binary/raw/v1_1/immutable_buffer.rs

```diff
@@ -275,76 +277,198 @@ impl<'a> ImmutableBuffer<'a> {
+                FlexSymValue::SymbolRef(_sym) => input.offset(),
+                FlexSymValue::Opcode(Opcode { opcode_type: OpcodeType::DelimitedContainerClose, ..}) => {
+                    offsets.push(after.offset() - 1);
+                    break
+                }
+                _ => unreachable!(),
+            };
+            input = after;
+
+            let mut opcode = input.peek_opcode()?;
+            if opcode.opcode_type == OpcodeType::Nop {
+                let res = input.consume_nop_padding(opcode)?;
+                input = res.1;
+                opcode = input.peek_opcode()?;
+            }
+            let value = input.read_value(opcode)?;
+            input = input.consume(value.encoded_value.total_length());
+            offsets.push(field_offset);
+        }
+
+        let header = head_opcode
+            .to_header()
+            .ok_or_else(|| IonError::decoding_error("found a non-value in value position"))?;
+        let value_body_length = *offsets.last().unwrap() - header_offset;
+        let total_length = 2 + value_body_length;
 
         let encoded_value = EncodedValue {
             header,
-            // If applicable, these are populated by the caller: `read_annotated_value()`
             annotations_header_length: 0,
             annotations_sequence_length: 0,
             annotations_encoding: AnnotationsEncoding::SymbolAddress,
             header_offset,
-            length_length,
-            value_body_length: value_length,
+            length_length: 0,
             total_length,
+            value_body_length,
         };
+
         let lazy_value = LazyRawBinaryValue_1_1 {
             encoded_value,
-            // If this value has a field ID or annotations, this will be replaced by the caller.
             input: self,
+            delimited_offsets: Some(offsets.into_bump_slice()),
         };
+
         Ok(lazy_value)
     }
 
+    /// Reads a value from the buffer. The caller must confirm that the buffer is not empty and that
+    /// the next byte (`type_descriptor`) is not a NOP.
+    pub fn read_value(self, opcode: Opcode) -> IonResult<LazyRawBinaryValue_1_1<'a>> {
+        if opcode.is_annotations_sequence() {
+            self.read_annotated_value(opcode)
+        } else {
+            self.read_value_without_annotations(opcode)
+        }
+    }
+
+    /// Reads a value from the buffer. The caller must confirm that the buffer is not empty and that
+    /// the next byte (`type_descriptor`) is neither a NOP nor an annotations wrapper.
+    fn read_value_without_annotations(
+        self,
+        opcode: Opcode,
+    ) -> IonResult<LazyRawBinaryValue_1_1<'a>> {
+        let input = self;
+        let header = opcode
+            .to_header()
+            .ok_or_else(|| IonError::decoding_error("found a non-value in value position .."))?;
+
+        if opcode.is_delimited() {
+            self.peek_delimited_container(opcode)
+        } else {
```

Comment:
The `peek_delimited_sequence` and `peek_delimited_struct` methods reproduce a lot of the code below this point. Would it be possible to refactor them to return their expression cache and the length consumed so they could still share the `EncodedValue` construction code in this method?

Reply:
Missed this comment in the previous commit, following up quick to address it.